### ISSUE TITLE
Add support for Dynamic Bucket Provisioning

### ISSUE THREAD

**Issue by bassam** [rook/rook#1705]
**Feature Request

With Minio and Ceph RGW support now in the Rook project, we need to enable a dynamic bucket provisioning similar to dynamic volume provisioning in Kubernetes. The goal is that application developers need not care about which service (managed services or in-cluster implementation) is providing the bucket for them. Cluster admins should be able to specify a BucketClass that includes details of the implementation.

Initial support should include S3, Minio and Ceph.

Design document coming /cc @tonya11en

**Comment by dgonzalezruiz** [comment]
I would see more value in creating users for S3/Swift. Also, at least in the case of RGWs, the buckets need to be owned by an UID. Please let me know if I can provide more details on the usecase.

**Comment by bassam** [comment]
@dgonzalezruiz yes please! :-)

**Comment by travisn** [comment]
@dgonzalezruiz #1583 should be what you want and definitely needed!

**Comment by dgonzalezruiz** [comment]
@bassam @travisn Sorry for the delay! I will point to my personal usecase in #1583. Thanks Travis!

**Comment by yehudasa** [comment]
@tonya11en @bassam Bucket provisioning for RGW via kubernetes service broker:
https://github.com/yehudasa/rgw-object-broker

Might be somewhat related.

**Comment by tonya11en** [comment]
@yehudasa the goals between the two are different. To consume buckets via the rgw-object-broker, the user needs to know the `bucket name, the RGW endpoint, and the credentials` according to the documentation. We cannot accomplish "write once, run anywhere" portability with the object broker this way since it requires the user to have knowledge about the Ceph cluster in the environment the application will be running.

In the proposed design, an admin must create a storage/bucket class for the Rook cluster's object storage (either Minio or Ceph) and the user just needs to know the details of the bucket they want to consume. No need for understanding what object storage providers are in the environment or the endpoint details.

**Comment by yehudasa** [comment]
@tonya11en not quite.  Sorry if the docs were misleading, I may need to make them a bit clearer. The admin that sets up the system does need to know the RGW endpoint and have an admin user credentials ahead of time. A user of the system (e.g., a client app) doesn't. It just needs to know the name of the broker (e.g., rgw-bucketbroker). Then it can list the existing plans (currently there's only one), create an instance (the user provides a name for the instance, and *optionally* a bucket name), bind to the instance (where it will receive an endpoint, bucket name, and credentials to use in order to access the bucket: this is needed for any solution, a user cannot access the bucket through the RESTful api without at least knowing that). Other users can also bind, each will get newly created unique sets of credentials. Once done the user can the unbind (which will void its credentials), and at the end it can remove the instance (suspend bucket, data will not be accessible anymore, link it it to a garbage collection user).

**Comment by liewegas** [comment]
If we go with a service broker-based approach, presumably that first part (configuring the broker itself) would be done by rook automagically.

I'm by no means an expert on kubernetes or service brokers, but it seems to me like the real question here is whether the user experience for magically getting the bucket is a rook-invented workload based on rook CRDs or whether it aligns with the service broker approach.  I know that kubernetes has "adopted" the service broker model in some sense, but I don't have any sense of how widely it is being embraced etc.  I do worry about rook inventing a different way just because when the server broker way is available and gets the jobs done just as well.

**Comment by yehudasa** [comment]
In any case, whatever the direction and decision you make, keep in mind that the service broker code exists and works and provides the functionality that is needed for RGW. It can probably be used for minio with some modification (the RGW specific stuff deals with RGW user management).

**Comment by bassam** [comment]
At a high level it seems that both approaches are trying to achieve similar goals -- make it easy for users to consume buckets from pods. It would be great if we can collaborate on a common solution.

Rook's approach is modeled after PersistentVolumes and StorageClass APIs in Kubernetes proper. At a high level I'm not sure why we would even expose the user with the concept of a "broker" when what they need is a bucket. It's almost like saying you need a Volume Broker (RBD, EBS, Google PD, etc.) to consume a volume from a pod. As @liewegas Rook could implement this by using a broker itself behind the scenes.

We don't want to reinvent the wheel here, but we do think that aligning more closely to the "Volume" design provides a more friendly and portable experience.

**Comment by yehudasa** [comment]
@bassam while not necessary, a user can be used when setting ACLs on objects.

**Comment by allen13** [comment]
In combination with the CephObjectStoreUser CRD it seems like it would be pretty straight forward to create something like a CephObjectStoreBucket CRD using https://github.com/minio/minio-go. Is anyone still planning on implementing this?

**Comment by liewegas** [comment]
There is an project ongoing to prototype a PV-like workflow to do this.  Basic idea is that a StorageClass is associated with an particular object store (e.g., RGW), (or, in the future, perhaps an S3 region), app  create a ObjectStorageUserClaim and ...BucketClaim, rook goes and creates the user/bucket and populates secrets, and then the app can mount the secrets (or whatever) to access the buckets.  I think the plan is to present current status shortly in one of the next community calls.

**Comment by bassam** [comment]
There’s also work that we merged in crossplane that is related https://github.com/crossplaneio/crossplane/pull/226 would be good to discuss how these all come together. 

This document describes our approach which is also inspired by PV:
https://docs.google.com/document/d/1whncqdUeU2cATGEJhHvzXWC9xdK29Er45NJeoemxebo
